README.md
# Gnome Extension - Custom Top Bar Menu
Customizable menu on Gnome top bar.

![Screenshot Preview](screenshot.png)
![another Preview](widget.png)


### Usage:
- A `.custom-top-bar-menu-entries.json` file will automatically be created within your $HOME directory, if it does not already exist. (**_$HOME/.custom-top-bar-menu-entries.json_**).

- Edit the file by clicking [Edit Config File] within the Custom Menu Config submenu and add your favorite commands, referencing the '.custom-top-bar-menu-entries.json' file in this examples directory for ideas. It is a `.json` file so respect syntax and commands accordingly. For example, use commas after each item, but not on the last item.

- Reload the configuration file for changes to take effect by choosing *[Reload Configuration File]* from the *Custom Menu Config* submenu. Alternatively, disable/enable the plugin within the Extensions app or restart the entire Gnome environment using (Alt+F2, 'r', Enter).

- If you make a mistake in the '.custom-top-bar-menu-entries.json' file and a JSON parse error occurs, a submenu entitled *'Config File Parse Error . . . '* is displayed in the custom menu and automatically expands to show *[Edit Config File]*, *[Reload Config File]* and *[View Log in Journalctl]*.

- The optional constants, 'editorExecutable', defaulted to 'gedit' and 'terminalExecutable', defaulted to 'gnome-terminal' are currently only used for the permanent *Custom Menu Config* submenu items , where *[Edit Config File]* uses 'editorExecutable' and *[View Log in Journalctl]* uses 'terminalExecutable'. These have been tested with Geany and Tilix, respectively.

The widgets manual below describe available entry types, and the '.custom-top-bar-menu-entries.json' file in the `examples` directory showcases usage thereof.


---

## Widgets Manual
Here are common widgets and entities you can use with this plugin:
### **- Launcher**
Create a new entry and put it in the menu
```json
    {
      "type": "launcher",
      "title": "Item Name on Menu",
      "command": "/your/command/to/execute --with-parameters"
    },
```

### **- Separator**
Place a menu separator _(\<hr>)_
```json
    {
      "type": "separator"
    },
```

### **- SubMenu**
Create a submenu inside current menu, items inside the submenu are placed inside `entries`
and they can be of any type (launcher, separator, submenu, ...)
```json
    {
        "type": "submenu",
        "title": "Menu Name",
        "entries": [
            //... Place your entries here ...//
        ]
    },
```

### **- Toggler**
Create a toggle item on menu, it has a detector and ON|OFF commands.
Command detector can be _activated|deactivated_ by a **0** _(success)_ return error level
and later detected with a plain JavaScript eval. See graphical result on above screenshots
```json
    {
      "type": "toggler",
      "title":       "Widget Title",
      "command_on":  "/command/when/turned/on",
      "command_off": "/command/when/turned/off",
      "detector":    "/command/detector > /dev/null && echo yes"
    },
```


## History and License
License: GPL v3

This expands upon previous GPL works listed below in the predecessors section by adding:
- Expansion of Gnome compatibility to v43
- The ability to reload the configuration file, rebuild the menu and revalidate the configuration file JSON syntax without having to restart Gnome using the "Alt-F2, r" method or by logging in/out of the Gnome session.
- Added optional constants to the default configuration file (~/.custom-top-bar-menu-entries.json)
	- 'editorExecutable', defaulted to 'gedit' (for Gnome < 42), 'gnome-text-editor' (for Gnome > 43) and
	- 'terminalExecutable', defaulted to 'gnome-terminal'.
	- These values are currently only used for the permanent 'Custom Menu Config' submenu items, where [Edit Config File] uses 'editorExecutable' and [View Log in Journalctl] uses 'terminalExecutable'.
	- The user remains able to specify their preferred text editor in the confiugration file to deviate from the default.
		- Tested custom values for 'editorExecutable' with Geany and 'terminalExecutable' with Tilix.
	- If a configuration file parsing error is detected, the appropriate text-editor is determined based on the Gnome Version.
	- As gnome-text-editor has replaced gedit in Gnome 43, the default text editor is now defined as a function of Gnome version being >= 43 or less than 43. The default text editor is used during generation of a default configuration file, as well as used as an editor to spawn when the configuration file parsing error is detected. 
- Added automatic expansion of the 'Custom Menu Config' submenu if ~/.entires JSON parsing fails.
---

GPL Predecessors (Newer to Older)

- https://github.com/andreabenini/gnome-plugin.custom-menu-panel (GPL v3)
- https://github.com/Shihira/gnome-extension-quicktoggler (GPL v2)
	- Deprecated - Author switched to developing for KDE.


CHANGELOG.txt
2022-10-14 v5.1
				- Gnome 43 compatibility added.
				- As gnome-text-editor has replaced gedit in Gnome 43, the default text editor is now defined as a function of Gnome version being >= 43 or less than 43. The derfault text editor is used during generation of a default configuration file, as well as used as an editor to spawn when the configuration file parsing error is detected. The user remains able to specify their preferred text editor in the confiugration file to deviate from the default.
				- Renamed extension to custom-top-bar-menu from custom-menu-panel

2022-09-24 v5
				Resolved issues:
					- Default configuration file would overwrite the current ~/.entries file if any syntax error was detected during JSON parsing. The configuration file is now only overwritten if no file is currently present. If a JSON parsing error is discovered, a submenu entitled 'Config File Parse Error . . . ' is displayed in the custom menu and automatically expanded to show [Edit Config File], [Reload Config File] and [View Log in Journalctl].
				Added capabilities:
					- Added the ability to reload the configuration file, rebuilding the menu and revalidating the JSON syntax without having to restart Gnome using the "Alt-F2, r" method.
					- Added 'Custom Menu Config' submenu to the end of the menu, regardless of the state of ~/.entries, containing [Edit Config File], [Reload Config File] and [View Log in Journalctl].
					- Added optional constants in ~/.entries 'editorExecutable', defaulted to 'gedit' and 'terminalExecutable', defaulted to 'gnome-terminal'. These values are currently only used for the permanent 'Custom Menu Config' submenu items, where [Edit Config File] uses 'editorExecutable' and [View Log in Journalctl] uses 'terminalExecutable'. Tested 'editorExecutable' with Geany. Tested 'terminalExecutable' with Tilix.
					- Added automatic expansion of the 'Custom Menu Config' submenu if ~/.entires JSON parsing fails.
					
2022-04 v4.1.0
        - Gnome 42 update

2022-03 v4.0.0
        - New widget: status toggler. Detects and toggle on/off the widget with your favorite command

2021-07 v3.0.0
        - Gnome 40 and 41 fixes

2021-06 v2.0.0
        - Project cleanup and refactoring

2021-05 v1.0.1
        - Bugfix. On startup it creates $HOME/.entries.json file when it does not exists for later editing

2021-05 v1.0.0
        - Note.  First version published on gnome extensions portal, requires .entries.json to work properly



custom-top-bar-menu@rokkoyama/extension.js
/* extension.js
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

/**
 * @author Rokkoyama
 * @see    https://github.com/rokkoyama/gnome-plugin.custom-menu-panel/
 */

/* exported init */

const GETTEXT_DOMAIN = 'custom-top-bar-menu';
const CONFIGURATION_FILE = '.custom-top-bar-menu-entries.json';

const { GObject, St } = imports.gi;

const Gettext = imports.gettext.domain(GETTEXT_DOMAIN);
const _       = Gettext.gettext;

const ExtensionUtils    = imports.misc.extensionUtils;
const Lang              = imports.lang;
const GLib              = imports.gi.GLib;
const Gio               = imports.gi.Gio;
const Main              = imports.ui.main;
const PanelMenu         = imports.ui.panelMenu;
const PopupMenu         = imports.ui.popupMenu;

const Me                = imports.misc.extensionUtils.getCurrentExtension();
const Config            = Me.imports.config

const LOGGER_INFO       = 0;
const LOGGER_WARNING    = 1;
const LOGGER_ERROR      = 2;

const Indicator = GObject.registerClass(
	class Indicator extends PanelMenu.Button {
		_init() {
				super._init(0.0, _('Custom Menu Panel Indicator'));
				this.add_child(new St.Icon({
						icon_name: 'view-list-bullet-symbolic',
						style_class: 'system-status-icon',
				}));
				this._loadSetup();
		}

		/**
		 * LOAD Program settings from .entries.json file
		 */
		_loadSetup() {
				this.menu.removeAll();
				this.configLoader = new Config.Loader();
				// Check if config file does NOT exist in home directory
				let configFileExistenceTestResult = GLib.file_test(GLib.get_home_dir() + '/' + CONFIGURATION_FILE, GLib.FileTest.EXISTS);
				if (configFileExistenceTestResult == 0) {
					log(`${Me.metadata.name}: Config File does not exist at path: ` + GLib.get_home_dir() + '/' + CONFIGURATION_FILE);
					log(`${Me.metadata.name}: Creating default configuration file`);
					try { // Create a configuration file with default entries
						this.configLoader.saveDefaultConfig(GLib.get_home_dir() + '/' + CONFIGURATION_FILE);
					}
					catch(e) {
						logError(e,`${Me.metadata.name}: Error occurred while creating default config file at: ` + GLib.get_home_dir() + '/' + CONFIGURATION_FILE);
					}
				}
				else { // Config file found
					log(`${Me.metadata.name}: Config File Found at: ` + GLib.get_home_dir() + '/' + CONFIGURATION_FILE);
				}
				try { // Test config file for parsing errors
					var configErr = 0;
					this.configLoader.loadConfig(GLib.get_home_dir() + '/' + CONFIGURATION_FILE); // $HOME/.entries.json
				} catch(e) {
					logError(e,`${Me.metadata.name}: Config File Parse Error. Edit Config File and Inspect Syntax. Path: ` + GLib.get_home_dir() + '/' + CONFIGURATION_FILE);
					configErr = 1;
				}
				if (configErr == 0) { // Config file parse successful. Build the menu based on the configuration file, and add a submenu for configuration editing, reloading the config file and viewing the log.
					let i = 0;
					for (let i in this.configLoader.entries) {
							let item = this.configLoader.entries[i].createItem();
							this.menu.addMenuItem(item);
					}
					this.menu.addMenuItem(new PopupMenu.PopupSeparatorMenuItem());
					let subMenu = new PopupMenu.PopupSubMenuMenuItem('Custom Menu Config');
					this.menu.addMenuItem(subMenu);
					subMenu.menu.addAction('[Edit Config File]', e => this.EditConfigFile());
					subMenu.menu.addAction('[Reload Configuration File]', e => this._loadSetup())
					subMenu.menu.addAction('[View Log in Journalctl]', e => this.ViewGnomeLog());
				}
				else // Config file parse error. Display 'Config File Parse Error . . .' as the title of a submenu containing items for configuration editing, reloading the config file and viewing the log.
				{
					let subMenu = new PopupMenu.PopupSubMenuMenuItem('Config File Parse Error . . .');
					this.menu.addMenuItem(subMenu);
					subMenu.menu.addAction('[Edit Config File]', e => this.EditConfigFile());
					subMenu.menu.addAction('[Reload Configuration File]', e => this._loadSetup())
					subMenu.menu.addAction('[View Log in Journalctl]', e => this.ViewGnomeLog());
					this.menu.connect('open-state-changed', Lang.bind(this, function() {
						subMenu.setSubmenuShown(true);
					}));
				}
		}
		ViewGnomeLog() {
			let termArg = '--';
			if (this.configLoader.terminalExecutable == "tilix") {
				termArg = '-e';
			}
			const command = this.configLoader.terminalExecutable + ' ' + termArg + ' ' + "bash -i -c \'journalctl -f -o cat /usr/bin/gnome-shell\'";
			try {
				Config._generalSpawn(command, this.configLoader.__env, 'Gnome-Shell Log');
			}
			catch (e) {
				logError(e,`${Me.metadata.name}: Viewing Gnome-Shell Log in ${terminalExecutable} failed`);
			}
		}
		 EditConfigFile() {
			//const command = [this.configLoader.editorExecutable, GLib.get_home_dir() + '/' + CONFIGURATION_FILE];
			const command = this.configLoader.editorExecutable + ' ' + GLib.get_home_dir() + '/' + CONFIGURATION_FILE;
			try {
				Config._generalSpawn(command, this.configLoader.__env, 'Edit Custom Menu Config File');
			}
			catch (e) {
				logError(e,`${Me.metadata.name}: Attempt to edit the configuration file in ${editorExecutable} failed`);
			}
		}
	}
);


const Logger = new Lang.Class({
    Name: 'Logger',

    _init: function(log_file) {
        this._log_file = log_file;
        // initailize log_backend
        if(!log_file)
            this._initEmptyLog();
        else if(log_file == "gnome-shell")
            this._initGnomeLog();
        else
            this._initFileLog();

        this.level = LOGGER_WARNING;

        this.info = function(t) {
            if(this.level <= LOGGER_INFO) this.log(t)
        };
        this.warning = function(t) {
            if(this.level <= LOGGER_WARNING) this.log(t)
        };
        this.error = function(t) {
            if(this.level <= LOGGER_ERROR) this.log(t);
        };
    },

    _initEmptyLog: function() {
        this.log = function(_) { };
    },

    _initGnomeLog: function() {
        this.log = function(s) {
            global.log("custom-menu-panel> " + s);
        };
    },

    _initFileLog: function() {
        this.log = function(s) {
            // all operations are synchronous: any needs to optimize?
            if(!this._output_file || !this._output_file.query_exists(null) ||
                !this._fstream || this._fstream.is_closed()) {

                this._output_file = Gio.File.new_for_path(this._log_file);
                this._fstream = this._output_file.append_to(
                    Gio.FileCreateFlags.NONE, null);

                if(!this._fstream instanceof Gio.FileIOStream) {
                    this._initGnomeLog();
                    this.log("IOError: Failed to append to " + this._log_file +
                            " [Gio.IOErrorEnum:" + this._fstream + "]");
                    return;
                }
            }

            this._fstream.write(String(new Date())+" "+s+"\n", null);
            this._fstream.flush(null);
        }
    },

    notify: function(t, str, details) {
        this.ncond = this.ncond || ['proc', 'ext', 'state'];
        if(this.ncond.indexOf(t) < 0) return;
        Main.notify(str, details || "");
    },
});

// lazy-evaluation
let logger = null;
function getLogger() {
    if(logger === null)
        logger = new Logger("gnome-shell");
    return logger;
}

class Extension {
    constructor(uuid) {
        this._uuid = uuid;

        ExtensionUtils.initTranslations(GETTEXT_DOMAIN);
    }

    enable() {
        this._indicator = new Indicator();
        Main.panel.addToStatusArea(this._uuid, this._indicator);
    }

    disable() {
        this._indicator.destroy();
        this._indicator = null;
    }
}

function init(meta) {
    return new Extension(meta.uuid);
}


custom-top-bar-menu@rokkoyama/config.js
const Gio  = imports.gi.Gio;
const GLib = imports.gi.GLib;
const Json = imports.gi.Json;
const Lang = imports.lang;
const Main = imports.ui.main;
const PopupMenu = imports.ui.popupMenu;
const ByteArray = imports.byteArray;

const Me = imports.misc.extensionUtils.getCurrentExtension();
const getLogger = Me.imports.extension.getLogger;
const Config = imports.misc.config;

const Entry = new Lang.Class({
    Name: 'Entry',
    Abstract: true,

    _init: function(prop) {
        this.type = prop.type;
        this.title = prop.title || "";

        this.__vars = prop.__vars || [];
        this.updateEnv(prop);
    },

    setTitle: function(text) {
        this.item.label.get_clutter_text().set_text(text);
    },

    updateEnv: function(prop) {
        this.__env = {}
        if(!this.__vars) return;

        for(let i in this.__vars) {
            let v = this.__vars[i];
            this.__env[v] = prop[v] ? String(prop[v]) : "";
        }
    },

    // the pulse function should be read as "a pulse arrives"
    pulse: function() { },

    _try_destroy: function() {
        try {
            if(this.item && this.item.destroy)
                this.item.destroy();
        } catch(e) { /* Ignore all errors during destory*/ }
    },
});

const DerivedEntry = new Lang.Class({
    Name: 'DerivedEntry',

    _init: function(prop) {
        if(!prop.base)
            throw new Error("Base entry not specified in type definition.");

        this.base = prop.base;
        this.vars = prop.vars || [];

        delete prop.base;
        delete prop.vars;

        this.prop = prop;
    },

    createInstance: function(addit_prop) {
        let cls = type_map[this.base];
        if(!cls) throw new Error("Bad base class.");
        if(cls.createInstance) throw new Error("Not allowed to derive from derived types");

        for(let rp in this.prop)
            addit_prop[rp] = this.prop[rp];
        addit_prop.__vars = this.vars;

        let instance = new cls(addit_prop);

        return instance;
    },
});

/*
 * callback: function (stdout, stderr, exit_status) { }
 */
let __pipeOpenQueue = [];
let __pipeExecTimer = null;

function pipeOpen(cmdline, env, callback) {
    let param = [cmdline, env, callback]
    __pipeOpenQueue.push(param);
    if(__pipeExecTimer === null) {
        __pipeExecTimer = GLib.timeout_add(GLib.PRIORITY_DEFAULT, 50,
            function() {
            let param = __pipeOpenQueue.shift();
            if(param === undefined) {
                __pipeExecTimer = null;
                return false;
            }
            if(realPipeOpen) realPipeOpen(param[0], param[1], param[2]);
            return true;
        });

    }
}

function realPipeOpen(cmdline, env, callback) {
    let user_cb = callback;
    let proc;

    function wait_cb(_, _res) {
        let stdout_pipe = proc.get_stdout_pipe();
        let stderr_pipe = proc.get_stderr_pipe();

        let stdout_content;
        let stderr_content;

        // Only the first GLib.MAXINT16 characters are fetched for optimization.
        stdout_pipe.read_bytes_async(GLib.MAXINT16, 0, null, function(osrc, ores) {
            stdout_content = ByteArray.toString(stdout_pipe.read_bytes_finish(ores).get_data());
            stdout_pipe.close(null);

            stderr_pipe.read_bytes_async(GLib.MAXINT16, 0, null, function(esrc, eres) {
                stderr_content = ByteArray.toString(stderr_pipe.read_bytes_finish(eres).get_data());
                stderr_pipe.close(null);

                user_cb(stdout_content, stderr_content, proc.get_exit_status());
            });
        });
    }

    if(user_cb) {
        let _pipedLauncher = new Gio.SubprocessLauncher({
            flags:
                Gio.SubprocessFlags.STDERR_PIPE |
                Gio.SubprocessFlags.STDOUT_PIPE
        });
        for(let key in env) {
            _pipedLauncher.setenv(key, env[key], true);
        }
        proc = _pipedLauncher.spawnv(['bash', '-c', cmdline]);
        proc.wait_async(null, wait_cb);
    } else {
        // Detached launcher is used to spawn commands that we are not concerned
        // about its result.
        let _detacLauncher = new Gio.SubprocessLauncher();
        for(let key in env) {
            _detacLauncher.setenv(key, env[key], true);
        }
        proc = _detacLauncher.spawnv(['bash', '-c', cmdline]);
    }

    getLogger().info("Spawned " + cmdline);

    // log(`${Me.metadata.name}: Spawning ${cmdline}`);   TODO: BEN
    return proc.get_identifier();
}

function _generalSpawn(command, env, title) {
    title = title || "Process";
    pipeOpen(command, env, function(stdout, stderr, exit_status) {
        if(exit_status != 0) {
            log
            getLogger().warning(stderr);
            getLogger().notify("proc", title +
                " exited with status " + exit_status, stderr);
        }
    });
}

function quoteShellArg(arg) {
    arg = arg.replace(/'/g, "'\"'\"'");
    return "'" + arg + "'";
}

// This cache is used to reduce detector cost. Each time creating an item, it
// check if the result of this detector is cached, which prevent the togglers
// from running detector on each creation. This is useful especially in search
// mode.
let _toggler_state_cache = { };

const TogglerEntry = new Lang.Class({
    Name: 'TogglerEntry',
    Extends: Entry,

    _init: function(prop) {
        this.parent(prop);

        this.command_on = prop.command_on || "";
        this.command_off = prop.command_off || "";
        this.detector = prop.detector || "";
        this.auto_on = prop.auto_on || false;
        this.notify_when = prop.notify_when || [];
        // if the switch is manually turned off, auto_on is disabled.
        this._manually_switched_off = false;
    },

    createItem: function() {
        this._try_destroy();
	this.item = new PopupMenu.PopupSwitchMenuItem(this.title, false);
	this.item.label.get_clutter_text().set_use_markup(true);
	this.item.connect('toggled', Lang.bind(this, this._onManuallyToggled));
        this._loadState();
        return this.item;
    },

    _onManuallyToggled: function(_, state) {
        // when switched on again, this flag will get cleared.
        this._manually_switched_off = !state;
        this._storeState(state);
        this._onToggled(state);
    },

    _onToggled: function(state) {
        if(state)
            _generalSpawn(this.command_on, this.__env, this.title);
        else
            _generalSpawn(this.command_off, this.__env, this.title);
    },

    _detect: function(callback) {
        // abort detecting if detector is an empty string
        if(!this.detector)
            return;

        pipeOpen(this.detector, this.__env, function(out) {
            out = String(out);
            callback(!Boolean(out.match(/^\s*$/)));
        });
    },

    compareState: function(new_state) {
        // compare the new state with cached state
        // notify when state is different
        let old_state = _toggler_state_cache[this.detector];
        if(old_state === undefined) return;
        if(old_state == new_state) return;

        if(this.notify_when.indexOf(new_state ? "on" : "off") >= 0) {
            let not_str = this.title + (new_state ? " started." : " stopped.");
            if(!new_state && this.auto_on)
                not_str += " Attempt to restart it now.";
            getLogger().notify("state", not_str);
        }
    },

    _storeState: function(state) {
        let hash = JSON.stringify({ env: this.__env, detector: this.detector });
        _toggler_state_cache[hash] = state;
    },

    _loadState: function() {
        let hash = JSON.stringify({ env: this.__env, detector: this.detector });
        let state = _toggler_state_cache[hash]; 
        if(state !== undefined)
            this.item.setToggleState(state); // doesn't emit 'toggled'
    },

    pulse: function() {
        this._detect(Lang.bind(this, function(state) {
            this.compareState(state);

            this._storeState(state);
            this._loadState();
            //global.log(this.title + ': ' + this._manually_switched_off);

            if(!state && !this._manually_switched_off && this.auto_on)
                // do not call setToggleState here, because command_on may fail
                this._onToggled(this.item, true);
        }));
    },

    perform: function() {
        this.item.toggle();
    },
});

const LauncherEntry = new Lang.Class({
    Name: 'LauncherEntry',
    Extends: Entry,

    _init: function(prop) {
        this.parent(prop);

        this.command = prop.command || "";
    },

    createItem: function() {
        this._try_destroy();

        this.item = new PopupMenu.PopupMenuItem(this.title);
        this.item.label.get_clutter_text().set_use_markup(true);
        this.item.connect('activate', Lang.bind(this, this._onClicked));

        return this.item;
    },

    _onClicked: function(_) {
        _generalSpawn(this.command, this.__env, this.title);
    },

    perform: function() {
        this.item.emit('activate');
    },
});


const SubMenuEntry = new Lang.Class({
    Name: 'SubMenuEntry',
    Extends: Entry,

    _init: function(prop) {
        this.parent(prop)

        if(prop.entries == undefined)
            throw new Error("Expected entries provided in submenu entry.");

        this.entries = [];

        for(let i in prop.entries) {
            let entry_prop = prop.entries[i];
            let entry = createEntry(entry_prop);
            this.entries.push(entry);
        }
    },

    createItem: function() {
        this._try_destroy();

        this.item = new PopupMenu.PopupSubMenuMenuItem(this.title);
        this.item.label.get_clutter_text().set_use_markup(true);
        for(let i in this.entries) {
            let entry = this.entries[i];
            this.item.menu.addMenuItem(entry.createItem());
        }

        return this.item;
    },

    pulse: function() {
        for(let i in this.entries) {
            let entry = this.entries[i];
            entry.pulse();
        }
    }
});

const SeparatorEntry = new Lang.Class({
    Name: 'SeparatorEntry',
    Extends: Entry,

    _init: function(prop) { },

    createItem: function() {
        this._try_destroy();

        this.item = new PopupMenu.PopupSeparatorMenuItem(this.title);
        this.item.label.get_clutter_text().set_use_markup(true);

        return this.item;
    },
});

let type_map = {};

////////////////////////////////////////////////////////////////////////////////
// Config Loader loads config from JSON file.

// convert Json Nodes (GLib based) to native javascript value.
function convertJson(node) {
    if(node.get_node_type() == Json.NodeType.VALUE)
        return node.get_value();
    if(node.get_node_type() == Json.NodeType.OBJECT) {
        let obj = {}
        node.get_object().foreach_member(function(_, k, v_n) {
            obj[k] = convertJson(v_n);
        });
        return obj;
    }
    if(node.get_node_type() == Json.NodeType.ARRAY) {
        let arr = []
        node.get_array().foreach_element(function(_, i, elem) {
            arr.push(convertJson(elem));
        });
        return arr;
    }
    return null;
}

//
function createEntry(entry_prop) {
    if(!entry_prop.type)
        throw new Error("No type specified in entry.");

    let cls = type_map[entry_prop.type];
    if(!cls)
        throw new Error("Incorrect type '" + entry_prop.type + "'");
    else if(cls.createInstance)
        return cls.createInstance(entry_prop);

    return new cls(entry_prop);
}

var Loader = new Lang.Class({
    Name: 'ConfigLoader',

    _init: function(filename) {
				let gnomeVersion = Config.PACKAGE_VERSION.split('.')
				if (gnomeVersion[0] < 40 || gnomeVersion[0] > 43)
					throw new Error("Incompatible with Gnome Version "+gnomeVersion)
					
				if (gnomeVersion[0] >= 43) {
					// Default for Gnome 43 and subsequent, where gedit was removed and replaced with gnome-text-editor:
					this.editorExecutable = "gnome-text-editor";
				}
        else {
					this.editorExecutable = "gedit";
				}
				this.terminalExecutable = "gnome-terminal"; // Version-agnostic default.
				
        if(filename)
            this.loadConfig(filename);
    },

    loadConfig: function(filename) {
        // reset type_map everytime load the config
        type_map = {
            launcher: LauncherEntry,
            toggler: TogglerEntry,
            submenu: SubMenuEntry,
            separator: SeparatorEntry
        };

        type_map.systemd = new DerivedEntry({
            base: 'toggler',
            vars: ['unit'],
            command_on: "pkexec systemctl start ${unit}",
            command_off: "pkexec systemctl stop ${unit}",
            detector: "systemctl status ${unit} | grep Active:\\\\s\\*activ[ei]",
        });

        type_map.tmux = new DerivedEntry({
            base: 'toggler',
            vars: ['command', 'session'],
            command_on: 'tmux new -d -s ${session} bash -c "${command}"',
            command_off: 'tmux kill-session -t ${session}',
            detector: 'tmux has -t "${session}" 2>/dev/null && echo yes',
        });
				
        //Refer to README file for detailed config file format.
        this.entries = [];
        
        let config_parser = new Json.Parser();
        config_parser.load_from_file(filename);

        let conf = convertJson(config_parser.get_root());
        if (conf.entries == undefined)
            throw new Error("Key 'entries' not found.");
        if (conf.deftype) {
            for (let tname in conf.deftype) {
                if (type_map[tname])
                    throw new Error("Type \""+tname+"\" duplicated.");
                type_map[tname] = new DerivedEntry(conf.deftype[tname]);
            }
        }

        for (let conf_i in conf.entries) {
            let entry_prop = conf.entries[conf_i];
            this.entries.push(createEntry(entry_prop));
        }
       if (conf.editorExecutable != undefined) {
				 this.editorExecutable = conf.editorExecutable;
			 }
			 if (conf.terminalExecutable != undefined) {
        this.terminalExecutable = conf.terminalExecutable;
			}
    },


    saveDefaultConfig: function(filename) {
        // Write default config
        const PERMISSIONS_MODE = 0o640;
        const jsonString = JSON.stringify({
					"_homepage_": "https://github.com/rokkoyama/gnome-plugin.custom-top-bar-menu",
          "_examples_": "https://github.com/rokkoyama/gnome-plugin.custom-top-bar-menu/tree/main/examples",
          "editorExecutable": this.editorExecutable,
          "terminalExecutable": "gnome-terminal",
          "entries": [ {
						"type": "launcher",
						"title": "Example: List Home Directory in Terminal",
						"command": "gnome-terminal -- bash -i -c 'cd ~/; command ls -aF --color=auto; bash'"
					} ]
        }, null, 1);
        let fileConfig = Gio.File.new_for_path(filename);
        if (GLib.mkdir_with_parents(fileConfig.get_parent().get_path(), PERMISSIONS_MODE) === 0) {
            fileConfig.replace_contents(jsonString, null, false, Gio.FileCreateFlags.REPLACE_DESTINATION, null);
        }
        // Try to load newly saved file
        try {
            this.loadConfig(filename);
        } catch(e) {
            Main.notify(_('Cannot create and load file: '+filename));
        }
    }, /**/
		
});
